client/client.go

```diff
@@ -235,3 +253,132 @@ func (dfs *DistributedFileSystem) Read(path string, offset uint, length uint) ([
 
 	return dataBuffer, nil
 }
+
+type WriteHandle struct {
+	dfs    *DistributedFileSystem
+	path   string
+	offset uint
+}
+
+func (dfs *DistributedFileSystem) OpenForWrite(path string) WriteHandle {
+	return WriteHandle{
+		dfs:    dfs,
+		path:   path,
+		offset: 0,
+	}
+}
+
+func (writeHandle *WriteHandle) Seek(offset uint) {
+	writeHandle.offset = offset
+}
+
+func (writeHandle *WriteHandle) Write(data []byte) error {
+	offset := writeHandle.offset
+	length := uint(len(data))
+	path := writeHandle.path
+	dfs := writeHandle.dfs
+
+	if length == 0 {
+		return nil
+	}
+
+	// Step 1: Calculate the blocks that need to be read and their offsets
+	beginBlock := offset / common.BLOCK_SIZE
+	endBlock := (offset+length-1)/common.BLOCK_SIZE + 1
+	beginBlockOffset := offset % common.BLOCK_SIZE
+	endBlockOffset := (offset+length-1)%common.BLOCK_SIZE + 1
+
+	// Step 2: Get the block metadata from the NameNode
+	getBlockLocationsRequest := common.GetBlockLocationsRequest{
+		FileName:   path,
+		BeginBlock: uint(beginBlock),
+		EndBlock:   uint(endBlock),
+	}
+
+	var getBlockLocationsResponse common.GetBlockLocationsResponse
+	client, err := rpc.DialHTTP("tcp", dfs.endpoint)
+	if err != nil {
+		slog.Error("dialing error", "error", err)
+		return err
+	}
+
+	err = client.Call("NameNode.GetBlockLocations", getBlockLocationsRequest, &getBlockLocationsResponse)
+	if err != nil {
+		slog.Error("Error during GetBlockLocations Request", "error", err)
+		return err
+	}
+
+	slog.Info("GetBlockLocations succeeded", "file", path, "BlockInfoList", getBlockLocationsResponse.BlockInfoList)
+
+	// Step 3: Read the blocks from the DataNodes in parallel
+	var wg sync.WaitGroup
+	allSucceeded := true
+
+	for i, blockInfo := range getBlockLocationsResponse.BlockInfoList {
+		wg.Add(1)
+
+		beginOffset := uint(0)
+		endOffset := uint(common.BLOCK_SIZE)
+		if i == 0 {
+			beginOffset = uint(beginBlockOffset)
+		}
+		if i == len(getBlockLocationsResponse.BlockInfoList)-1 {
+			endOffset = uint(endBlockOffset)
+		}
+
+		go func(blockInfo common.BlockInfo, blockIndex uint, beginOffset uint, endOffset uint) {
+			defer wg.Done()
+			dataNodeEndpoint := blockInfo.DataNodeEndpoints[0]
+			dataNodeClient, err := rpc.DialHTTP("tcp", dataNodeEndpoint)
+			if err != nil {
+				slog.Error("dialing error", "error", err, "DataNode Endpoint", dataNodeEndpoint)
+				allSucceeded = false
+				return
+			}
+
+			writeBlockRequest := common.WriteBlockRequest{
+				FileName:         path,
+				BlockIndex:       blockIndex,
+				Version:          blockInfo.Version + 1,
+				BeginOffset:      beginOffset,
+				Data:             data[beginOffset+common.BLOCK_SIZE*blockIndex-writeHandle.offset : endOffset+common.BLOCK_SIZE*blockIndex-writeHandle.offset],
+				ReplicaEndpoints: blockInfo.DataNodeEndpoints,
+				IndexInChain:     0,
+			}
+
+			var unused bool
+			asyncRpcCall := dataNodeClient.Go("DataNode.WriteBlock", writeBlockRequest, &unused, nil)
+
+			writeBlockTimeout :=
```

Comment:
the readBlockTimeout is implemented in hyper_parameter.go, should we move these two in  the same place? (not sure which one is better)